src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -710,6 +806,114 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
   NaClAlignedFree(natp);
 }
 
+
+/**
+ * The following functions are used to reap any cages which have received a fatal signal.
+ * On launch, sel_main creates the Reaper thread which calls the fault teardown functions when
+ * a faulted thread is caught in the signal handler (nacl_signal.c)
+ * 
+ * The fault teardown function will cycle through all child threads in cage,
+ * tear them down, and call pthread_cancel upon them
+ * 
+ * Finally it will exit the parent thread and signal that the cage has exited with SIGKILL
+ * 
+ * Note: We implement it this way because
+ * 1. We can't fully teardown threads within the signal handler since
+ *  a. We can't unregister/unmap the handler while within itself
+ *  b. Most libc functions are invalid within the handler
+ * 2. We need to not only teardown the faulting thread itself, but any other threads that have been
+ *    launched wtihin that cage.
+ */
+
+
+void InitFatalThreadTeardown(void) {
+  if (!NaClMutexCtor(&teardown_mutex)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize handler cleanup mutex");
+  }
+  if (!NaClMutexCtor(&reapermut)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize handler cleanup mutex");
+  }
+  if (!NaClCondVarCtor(&reapercv)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize reaper cv");
+  }
+}
+
+void DestroyFatalThreadTeardown(void) {
+  NaClMutexDtor(&teardown_mutex);
+  NaClMutexDtor(&reapermut);
+  NaClCondVarDtor(&reapercv);
+}
+
+
+void AddToFatalThreadTeardown(struct NaClAppThread *natp) {
+    if (natp_to_teardown == natp) return;
+    NaClXMutexLock(&teardown_mutex);
+    NaClXMutexLock(&reapermut);
+    natp_to_teardown = natp;
+    natp->tearing_down = true;
+    NaClXCondVarBroadcast(&reapercv);
+    NaClXMutexUnlock(&reapermut);
+
+}
+
+void FatalThreadTeardown(void) {
+  struct NaClThread *thread;
+  int status = 137; // Fatal error signal SIGKILL
+
+    struct NaClApp *nap = natp_to_teardown->nap;
+
+    NaClXMutexLock(&nap->threads_mu);
+    int num_threads = NaClGetNumThreads(nap);
+
+    for(int i = 0; i < num_threads; i++) {
+
+      struct NaClAppThread *natp_child = NaClGetThreadMu(nap, i);
+      if (natp_child && natp_child != natp_to_teardown) {
+        struct NaClThread *child_thread;
+        child_thread = &natp_child->host_thread;
+        NaClAppThreadTeardownInner(natp_child, false);
+        NaClThreadCancel(child_thread);
+      }
+    }
+    
+    lind_exit(status, nap->cage_id);
+    (void) NaClReportExitStatus(nap, NACL_ABI_W_EXITCODE(status, 0));
+    thread = &natp_to_teardown->host_thread;
+    NaClXMutexUnlock(&nap->threads_mu);
+
+    NaClAppThreadTeardownInner(natp_to_teardown, false);
+    NaClThreadCancel(thread);
+    natp_to_teardown = NULL;
+    NaClXMutexUnlock(&teardown_mutex);
+
+}
+
+void ThreadReaper(void* arg) {
+  while (reap) {
+    NaClXMutexLock(&reapermut);
+
+    NaClXCondVarWait(&reapercv, &reapermut);
+    if (reap) FatalThreadTeardown();
+
+    NaClXMutexUnlock(&reapermut);
+  }
```

Comment:
Is there a reason to have the locking/unlocking in the loop, rather than outside? You get a signal when reap is set to false, so this wouldn't change anything really.